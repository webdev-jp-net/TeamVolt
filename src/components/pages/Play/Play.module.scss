@use 'style/_variable' as *;

.play {
  display: flex;
  flex-direction: column;

  * {
    user-select: none;
  }
}

.container {
  padding: rem(24);
}

.memberList {
  display: flex;
  gap: rem(24);
  justify-content: center;
  margin: rem(48) 0;
}

.member {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
}

.memberIcon {
  @include rect(rem(36));

  display: block;
  background-color: currentColor;
  border-radius: 50%;
}

.current {
  position: absolute;
  top: 100%;
  margin: rem(6) 0;
  font-size: rem(12);
}

.winner {
  position: absolute;
  bottom: 100%;
  margin: rem(6) 0;
  font-size: rem(12);
}

.visualizer {
  position: absolute;
  top: rem(120);
  right: rem(24);
  left: rem(24);
  display: grid;
  align-items: end;
}

.stock {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  gap: rem(24) 0;
  align-items: end;
  height: 1;
  margin: rem(12) 0;
  font-size: rem(72);
}

.stockItem {
  width: rem(72);
  line-height: 1;
}

.gauge {
  width: rem(36);
  height: rem(60);
  margin: 0 rem(12);

  &.--invisible {
    opacity: 0;
  }
}

.header {
  display: flex;
  gap: rem(18) rem(12);
  padding: rem(24);
}

.time {
  width: 4em;
}

// tapエフェクト
// 円が広がりながらフェードアウトするキーフレームアニメーション
@keyframes tap {
  0% {
    opacity: 1;
    transform: scale(0);
  }

  100% {
    opacity: 0;
    transform: scale(1);
  }
}

.tap {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  color: $c_dark;
  background: transparent;
  border: none;
  -webkit-tap-highlight-color: transparent;
}

.effect {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: -1;
  display: block;
  width: rem(36);
  height: rem(36);
  margin: auto;
  pointer-events: none;
  content: '';
  user-select: none;
  background-color: #f4981f80;
  border-radius: 50%;
  opacity: 0;

  .--tap & {
    animation: tap 0.3s ease-out;
  }
}

.result {
  display: grid;
  place-items: center;
  height: 100%;
  font-size: rem(60);
}

.menu {
  display: grid;
  gap: rem(18) rem(12);
  padding: rem(24);
  margin: auto 0 0;
}
