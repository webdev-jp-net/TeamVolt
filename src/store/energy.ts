import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import { doc, getDoc, updateDoc, arrayUnion, runTransaction } from 'firebase/firestore';
import { db } from 'firebaseDB';

type ChargeUnitsData = { member: string; count: number };

type State = {
  genEnergy: number;
  chargeUnits: ChargeUnitsData[];
};
const initialState: State = {
  genEnergy: 0,
  chargeUnits: [],
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview

export const energyGetApi = createApi({
  reducerPath: 'energyGetApi',
  baseQuery: async id => {
    if (id) {
      const docRef = doc(db, 'team', id);
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: docSnap.data().chargeUnits };
      } else {
        return { error: 'No such document!' };
      }
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // 獲得バッテリー取得
    getChargeUnits: builder.query<ChargeUnitsData[], string>({
      query: id => id,
    }),
  }),
});
export const { useGetChargeUnitsQuery } = energyGetApi;

export const energyPostApi = createApi({
  reducerPath: 'energyPostApi',
  baseQuery: async ({ id, value }: { id?: string; value: ChargeUnitsData }) => {
    if (id && value) {
      const docRef = doc(db, 'team', id);
      try {
        await runTransaction(db, async transaction => {
          const sfDoc = await transaction.get(docRef);
          if (!sfDoc.exists()) {
            // eslint-disable-next-line no-throw-literal
            throw 'Document does not exist!';
          }
          const oldList = sfDoc.data().chargeUnits || [];
          const newList = [
            ...oldList.filter((item: ChargeUnitsData) => item.member !== value.member),
            value,
          ];
          transaction.update(docRef, { chargeUnits: newList });
        });
      } catch (e) {
        console.log('Transaction failure:', e);
      }
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: docSnap.data().chargeUnits };
      } else {
        return { error: 'No such document!' };
      }
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // 獲得バッテリー追加
    addChargeUnits: builder.mutation<ChargeUnitsData[], { id: string; value: ChargeUnitsData }>({
      query: ({ id, value }) => ({
        operationType: 'remove_challenger',
        id,
        value,
      }),
    }),
  }),
});

export const { useAddChargeUnitsMutation } = energyPostApi;

const energy = createSlice({
  name: 'energy',

  initialState,

  reducers: {
    // 獲得エネルギーを更新
    updateGenEnergy: (state, action: PayloadAction<number>) => {
      return {
        ...state,
        genEnergy: action.payload,
      };
    },
    // 獲得エネルギーをリセット
    resetEnergy: () => {
      return {
        genEnergy: 0,
        chargeUnits: [],
      };
    },
  },

  extraReducers: builder => {
    // 成功: 獲得バッテリー取得
    builder.addMatcher(
      energyGetApi.endpoints.getChargeUnits.matchFulfilled,
      (state, action: PayloadAction<ChargeUnitsData[]>) => {
        const result = { chargeUnits: action.payload };
        return {
          ...state,
          ...result,
        };
      }
    );
    // 成功: 獲得バッテリー追加
    builder.addMatcher(
      energyPostApi.endpoints.addChargeUnits.matchFulfilled,
      (state, action: PayloadAction<ChargeUnitsData[]>) => {
        const result = { chargeUnits: action.payload };
        return {
          ...state,
          ...result,
        };
      }
    );
  },
});

// Action Creator
export const { updateGenEnergy, resetEnergy } = energy.actions;

// Reducer
export default energy.reducer;
