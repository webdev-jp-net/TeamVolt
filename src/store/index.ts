import { combineReducers, configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';

import user, { statusGetApi, statusUpdateApi } from './user';

const reducer = combineReducers({
  user,
  [statusGetApi.reducerPath]: statusGetApi.reducer,
  [statusUpdateApi.reducerPath]: statusUpdateApi.reducer,
});

const middleware = getDefaultMiddleware({ serializableCheck: false }).concat(
  statusGetApi.middleware,
  statusUpdateApi.middleware
);

export const store = configureStore({ reducer, middleware });

export type RootState = ReturnType<typeof reducer>;
