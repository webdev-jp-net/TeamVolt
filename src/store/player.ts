import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import {
  doc,
  getDoc,
  getDocs,
  updateDoc,
  arrayRemove,
  arrayUnion,
  addDoc,
  deleteField,
  collection,
  runTransaction,
} from 'firebase/firestore';
import { db } from 'firebaseDB';

import type { PlayerArticleData } from 'types/player';
import type { TeamArticleData, ChargeUnitsData, MemberData } from 'types/team';

type State = {
  localId: string;
  player?: string;
  playerList: PlayerArticleData[];
  selectedTeam?: string;
  myTeam?: TeamArticleData;
  genEnergy: number;
};

const initialState: State = {
  localId: '',
  playerList: [],
  genEnergy: 0,
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview
// 読込用
export const playerGetApi = createApi({
  reducerPath: 'playerGetApi',
  baseQuery: async () => {
    const playerRef = collection(db, 'player');
    const playerSnapshot = await getDocs(playerRef);
    const result = playerSnapshot.docs.map(doc => {
      return { id: doc.id, ...doc.data() };
    });
    return { data: result };
  },
  endpoints: builder => ({
    // プレイヤー情報取得
    getPlayer: builder.query<PlayerArticleData[], void>({
      query: () => ({}),
    }),
  }),
});
export const { useGetPlayerQuery } = playerGetApi;

// チーム情報取得
export const teamArticleGetApi = createApi({
  reducerPath: 'teamArticleGetApi',
  baseQuery: async ({ id }) => {
    const docRef = doc(db, 'team', id);
    const docSnap = await getDoc(docRef);
    if (docSnap.exists()) {
      return { data: { id, ...docSnap.data() } };
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // チーム情報取得
    getTeamArticle: builder.query<TeamArticleData, string>({
      query: id => ({ id }),
    }),
  }),
});
export const { useGetTeamArticleQuery } = teamArticleGetApi;

// 書込用
export const playerAddApi = createApi({
  reducerPath: 'playerAddApi',
  baseQuery: async ({ value }) => {
    const playerRef = collection(db, 'player');
    const docRef = await addDoc(playerRef, {
      localId: value,
    });
    return { data: { id: docRef.id, localId: value } };
  },
  endpoints: builder => ({
    // プレイヤー情報追加
    addPlayer: builder.mutation<PlayerArticleData, string>({
      query: value => ({ value }),
    }),
  }),
});
export const { useAddPlayerMutation } = playerAddApi;

// チーム情報の更新

// チームメンバー編集
type teamPostApiProps = {
  operationType: string;
  id?: string;
  value?: string | number | ChargeUnitsData | MemberData;
};
export const teamPostApi = createApi({
  reducerPath: 'teamPostApi',
  baseQuery: async ({ operationType, id, value }: teamPostApiProps) => {
    if (operationType === 'add_member' && id) {
      // メンバーを追加
      const docRef = doc(db, 'team', id);
      await updateDoc(docRef, {
        member: arrayUnion(value),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'remove_member' && id) {
      const docRef = doc(db, 'team', id);
      // メンバーを削除
      await updateDoc(docRef, {
        member: arrayRemove(value),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'add_challenger' && id) {
      const docRef = doc(db, 'team', id);
      // 代表者を登録
      await updateDoc(docRef, {
        challenger: value,
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'remove_challenger' && id) {
      const docRef = doc(db, 'team', id);
      // 代表者を削除
      await updateDoc(docRef, {
        challenger: deleteField(),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'add_charge_units' && id && value) {
      // 獲得バッテリーを登録
      const docRef = doc(db, 'team', id);
      try {
        await runTransaction(db, async transaction => {
          const sfDoc = await transaction.get(docRef);
          if (!sfDoc.exists()) {
            // eslint-disable-next-line no-throw-literal
            throw 'Document does not exist!';
          }
          const oldList = sfDoc.data().chargeUnits || [];
          const newList = [
            ...oldList.filter(
              (item: ChargeUnitsData) =>
                item.member !== (value as unknown as ChargeUnitsData).member
            ),
            value,
          ];
          transaction.update(docRef, { chargeUnits: newList });
        });
      } catch (e) {
        console.log('Transaction failure:', e);
      }
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'update_used_units' && id) {
      const docRef = doc(db, 'team', id);
      // 救出ミッション試行回数を登録
      await updateDoc(docRef, {
        usedUnits: value,
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'update_current_positions' && id) {
      const docRef = doc(db, 'team', id);
      // 救出ミッション進捗を登録
      await updateDoc(docRef, {
        currentPosition: value,
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else if (operationType === 'close_mission' && id) {
      const docRef = doc(db, 'team', id);
      // メンバー、役割、プレイ結果を削除
      await updateDoc(docRef, {
        member: deleteField(),
        challenger: deleteField(),
        chargeUnits: deleteField(),
        currentPosition: deleteField(),
        usedUnits: deleteField(),
      });
      // レスポンスで返す情報を再取得
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        return { data: { id, ...docSnap.data() } };
      } else {
        return { error: 'No such document!' };
      }
    } else {
      return { error: 'No such document!' };
    }
  },
  endpoints: builder => ({
    // メンバー追加
    addMember: builder.mutation<TeamArticleData, { id: string; value: MemberData }>({
      query: ({ id, value }) => ({
        operationType: 'add_member',
        id,
        value,
      }),
    }),
    // メンバー削除
    removeMember: builder.mutation<TeamArticleData, { id: string; value: MemberData }>({
      query: ({ id, value }) => ({
        operationType: 'remove_member',
        id,
        value,
      }),
    }),
    // 抽選結果追加
    addChallenger: builder.mutation<TeamArticleData, { id: string; value: string }>({
      query: ({ id, value }) => ({
        operationType: 'add_challenger',
        id,
        value,
      }),
    }),
    // 抽選結果追加
    removeChallenger: builder.mutation<TeamArticleData, { id: string; value: string }>({
      query: ({ id, value }) => ({
        operationType: 'remove_challenger',
        id,
        value,
      }),
    }),
    // 獲得バッテリー追加
    addChargeUnits: builder.mutation<TeamArticleData, { id: string; value: ChargeUnitsData }>({
      query: ({ id, value }) => ({
        operationType: 'add_charge_units',
        id,
        value,
      }),
    }),
    // 救出ミッション試行数を更新
    updateUsedUnits: builder.mutation<TeamArticleData, { id: string; value: number }>({
      query: ({ id, value }) => ({
        operationType: 'update_used_units',
        id,
        value,
      }),
    }),
    // 救出ミッション進捗を更新
    updateCurrentPosition: builder.mutation<TeamArticleData, { id: string; value: number }>({
      query: ({ id, value }) => ({
        operationType: 'update_current_positions',
        id,
        value,
      }),
    }),
    // ミッション情報をクリア
    closeMission: builder.mutation<TeamArticleData, string>({
      query: id => ({
        operationType: 'close_mission',
        id,
      }),
    }),
  }),
});
export const {
  useAddMemberMutation,
  useRemoveMemberMutation,
  useAddChallengerMutation,
  useRemoveChallengerMutation,
  useAddChargeUnitsMutation,
  useUpdateUsedUnitsMutation,
  useUpdateCurrentPositionMutation,
  useCloseMissionMutation,
} = teamPostApi;

const player = createSlice({
  name: 'player',

  initialState,

  reducers: {
    // ローカルID更新
    updateLocalId: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        localId: action.payload,
      };
    },
    // 所属チーム更新
    updateTeam: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        selectedTeam: action.payload,
      };
    },
    // 所属チーム脱退
    escapeTeam: state => {
      return {
        ...state,
        selectedTeam: undefined,
      };
    },
    // 獲得エネルギーを更新
    updateGenEnergy: (state, action: PayloadAction<number>) => {
      return {
        ...state,
        genEnergy: action.payload,
      };
    },
    // 獲得エネルギーをリセット
    resetEnergy: store => {
      return {
        ...store,
        genEnergy: 0,
      };
    },
  },
  extraReducers: builder => {
    // 成功: プレイヤー情報取得
    builder.addMatcher(
      playerGetApi.endpoints.getPlayer.matchFulfilled,
      (state, action: PayloadAction<PlayerArticleData[]>) => {
        state.playerList = action.payload;
      }
    );
    // 成功: プレイヤー追加
    builder.addMatcher(
      playerAddApi.endpoints.addPlayer.matchFulfilled,
      (state, action: PayloadAction<PlayerArticleData>) => {
        const result = { playerList: [...state.playerList, action.payload] };
        return {
          ...state,
          ...result,
        };
      }
    );
    // 成功: チーム情報取得
    builder.addMatcher(
      teamArticleGetApi.endpoints.getTeamArticle.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: メンバー追加
    builder.addMatcher(
      teamPostApi.endpoints.addMember.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: メンバー削除
    builder.addMatcher(
      teamPostApi.endpoints.removeMember.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 代表者追加
    builder.addMatcher(
      teamPostApi.endpoints.addChallenger.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 代表者削除
    builder.addMatcher(
      teamPostApi.endpoints.removeChallenger.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 獲得バッテリー追加
    builder.addMatcher(
      teamPostApi.endpoints.addChargeUnits.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 救出ミッション試行数更新
    builder.addMatcher(
      teamPostApi.endpoints.updateUsedUnits.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: 救出ミッション進捗更新
    builder.addMatcher(
      teamPostApi.endpoints.updateCurrentPosition.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
    // 成功: ミッション情報クリア成功
    builder.addMatcher(
      teamPostApi.endpoints.closeMission.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        state.myTeam = action.payload;
      }
    );
  },
});

// Action Creator
export const { updateLocalId, updateTeam, escapeTeam, updateGenEnergy, resetEnergy } =
  player.actions;

// Reducer
export default player.reducer;
