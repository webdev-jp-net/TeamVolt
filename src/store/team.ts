import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { createApi } from '@reduxjs/toolkit/query/react';
import { getDocs, addDoc, collection } from 'firebase/firestore';
import { db } from 'firebaseDB';
import { TeamArticleData } from 'types/team';

type State = {
  teamList: TeamArticleData[];
};

const initialState: State = {
  teamList: [],
};

// RTK Queryの設定
// https://redux-toolkit.js.org/rtk-query/overview
// read
export const teamGetApi = createApi({
  reducerPath: 'teamGetApi',
  baseQuery: async ({ path }) => {
    const teamRef = collection(db, path);
    const teamSnapshot = await getDocs(teamRef);
    const result = teamSnapshot.docs.map(doc => {
      return { id: doc.id, ...doc.data() };
    });
    return { data: result };
  },
  endpoints: builder => ({
    // チーム情報取得
    getTeam: builder.query<TeamArticleData[], void>({
      query: () => ({ path: 'team' }),
    }),
  }),
});
export const { useGetTeamQuery } = teamGetApi;

// write
export const teamAddApi = createApi({
  reducerPath: 'teamAddApi',
  baseQuery: async ({ value }) => {
    const teamRef = collection(db, 'team');
    const docRef = await addDoc(teamRef, {
      name: value,
      member: [],
    });
    return { data: { id: docRef.id, name: value, member: [] } };
  },
  endpoints: builder => ({
    // チーム情報追加
    addTeam: builder.mutation<TeamArticleData, string>({
      query: value => ({ value }),
    }),
  }),
});
export const { useAddTeamMutation } = teamAddApi;

const team = createSlice({
  name: 'team',

  initialState,

  reducers: {
    updateTeamList: (state, action: PayloadAction<TeamArticleData[]>) => {
      return {
        ...state,
        teamList: action.payload,
      };
    },
  },

  extraReducers: builder => {
    // 成功: チーム情報取得
    builder.addMatcher(
      teamGetApi.endpoints.getTeam.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData[]>) => {
        state.teamList = action.payload;
      }
    );
    // 成功: チーム追加
    builder.addMatcher(
      teamAddApi.endpoints.addTeam.matchFulfilled,
      (state, action: PayloadAction<TeamArticleData>) => {
        const result = { teamList: [...state.teamList, action.payload] };
        return {
          ...state,
          ...result,
        };
      }
    );
  },
});

// Action Creator
export const { updateTeamList } = team.actions;

// Reducer
export default team.reducer;
