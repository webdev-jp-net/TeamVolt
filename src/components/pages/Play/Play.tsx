import { FC, useCallback, useEffect, useMemo } from 'react';

import { useNavigate } from 'react-router-dom';

import { useDispatch, useSelector } from 'react-redux';

import { Button } from 'components/parts/Button';
import { RootState } from 'store';
import {
  updateMemberList,
  resetDrawResult,
  useGetEntriesQuery,
  useGetDrawResultQuery,
  useAddEntriesMutation,
  useRemoveEntriesMutation,
  useUpdateDrawResultMutation,
} from 'store/user';

import styles from './Play.module.scss';

export const Play: FC = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { userId, memberList, drawResult } = useSelector((state: RootState) => state.user);

  // 入室メンバー取得
  const {
    isSuccess: getEntriesSuccess,
    isError: getEntriesError,
    refetch: getEntriesRefetch,
  } = useGetEntriesQuery();

  // 抽選結果取得
  const {
    isSuccess: getDrawResultSuccess,
    isError: getDrawResultError,
    refetch: getDrawResultRefetch,
  } = useGetDrawResultQuery();

  // 成功: 入室メンバー取得
  useEffect(() => {
    if (getEntriesSuccess) console.log({ getEntriesSuccess });
  }, [getEntriesSuccess]);

  // 成功: 抽選結果取得
  useEffect(() => {
    if (getDrawResultSuccess) console.log({ getDrawResultSuccess });
  }, [getDrawResultSuccess]);

  // エラー: 入室メンバー取得
  useEffect(() => {
    if (getEntriesError) console.error({ getEntriesError });
  }, [getEntriesError]);

  // エラー: 抽選結果取得
  useEffect(() => {
    if (getDrawResultError) console.error({ getDrawResultError });
  }, [getDrawResultError]);

  // 初回データ取得済フラグ: 入室メンバー情報と抽選結果を取得できている
  const initialLoaded = useMemo(() => {
    return getEntriesSuccess && getDrawResultSuccess;
  }, [getEntriesSuccess, getDrawResultSuccess]);

  // 締切済フラグ: 抽選が終了しており自分も入室していない
  const isClosed = useMemo(() => {
    return drawResult && !memberList.includes(userId);
  }, [drawResult, memberList, userId]);

  // 配列の要素からランダムに一つ引き当てる
  const getRandomElement = (list: string[]) => {
    const index = Math.floor(Math.random() * list.length);
    return list[index];
  };

  // 入室メンバーを追加する
  const [
    sendAddEntries, // mutation trigger
    { isLoading: entriesAddLoading }, // mutation state
  ] = useAddEntriesMutation();

  // 入室メンバーを削除する
  const [
    sendRemoveEntries, // mutation trigger
    { isLoading: entriesRemoveLoading }, // mutation state
  ] = useRemoveEntriesMutation();

  // 抽選結果を更新する
  const [
    sendDrawResult, // mutation trigger
    { isLoading: drawResultLoading, isSuccess: drawResultSuccess, isError: drawResultError }, // mutation state
  ] = useUpdateDrawResultMutation();

  // 抽選結果を更新成功したらrefetchする
  useEffect(() => {
    if (drawResultSuccess) getDrawResultRefetch();
  }, [drawResultSuccess, getDrawResultRefetch]);

  // 最新の入室メンバー情報を取得し、抽選実行を確認する
  const handleDraw = useCallback(async () => {
    // 逐次処理で、以下の処理を実行する
    // 1. 抽選結果を取得する
    // 2. 抽選結果が''であれば、入室メンバーを取得する
    // 3. 入室メンバーの数を含んだ確認ダイアログを表示する
    // 4. OKなら抽選を実行する
    // 5. NGなら何もしない
    await getDrawResultRefetch();
    if (!drawResult) {
      await getEntriesRefetch();
      if (memberList.length > 1) {
        if (
          window.confirm(
            memberList.length +
              ' people are currently entering. Do you want to raffle with these members?'
          )
        ) {
          // 抽選
          const result = getRandomElement(memberList);
          // 書き込み
          sendDrawResult(result);
        }
      } else {
        alert('There are not enough members to draw.');
      }
    }
  }, [getDrawResultRefetch, drawResult, getEntriesRefetch, memberList, sendDrawResult]);

  // 入室処理
  useEffect(() => {
    // userIdが割り付けられていることを確認し、処理を続行する
    if (userId && initialLoaded) {
      // 入室する
      const addMyself = () => {
        sendAddEntries(userId);

        // 入室メンバーに自分を追加する
        const newMemberList = [...memberList, userId];
        // 重複を解消する
        const uniqueMemberList: string[] = Array.from(new Set<string>(newMemberList));
        dispatch(updateMemberList(uniqueMemberList));
      };

      // 抽選前
      if (!drawResult) {
        addMyself();
      } else if (!memberList.length) {
        // 抽選結果があるのに誰も入室していない場合は、抽選結果をリセットする
        sendDrawResult('');
        dispatch(resetDrawResult());
        addMyself();
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [userId, initialLoaded]);

  // 退室処理
  const handleExit = useCallback(() => {
    // userIdが割り付けられていることを確認し、処理を続行する
    if (initialLoaded) {
      // 入室メンバーから自分を削除する
      sendRemoveEntries(userId);
      // トップページに戻る
      navigate('/');

      // メンバーリストを初期化する
      dispatch(updateMemberList([]));
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [initialLoaded, memberList]);

  // 再読み込み
  const handleReload = useCallback(() => {
    getEntriesRefetch();
    getDrawResultRefetch();
  }, [getEntriesRefetch, getDrawResultRefetch]);

  // 初回読み込み
  useEffect(() => {
    handleReload();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div className={`l-page ${styles.play}`}>
      <div className={styles.container}>
        {!isClosed ? (
          <>
            <p>winner: {drawResult}</p>
            <p>myself: {userId}</p>

            <div className={styles.memberList}>
              {memberList.map((value, index) => (
                <div key={index} className={styles.member}>
                  {value === drawResult && <span className={styles.winner}>winner</span>}
                  <span className={styles.memberIcon} style={{ color: `#${value}` }}></span>
                  {value === userId && <span className={styles.current}>myself</span>}
                </div>
              ))}
            </div>
          </>
        ) : (
          <p>
            This draw is closed.
            <br />
            Please wait until the next draw.
          </p>
        )}
        <div className={styles.menu}>
          {!isClosed && (
            <>
              <Button
                handleClick={handleReload}
                disabled={entriesAddLoading || entriesRemoveLoading || !!drawResult}
              >
                reload
              </Button>
              <Button handleClick={handleDraw} disabled={drawResultLoading || !!drawResult}>
                drew
              </Button>
            </>
          )}
          <Button handleClick={handleExit}>exit</Button>
        </div>
      </div>
    </div>
  );
};
