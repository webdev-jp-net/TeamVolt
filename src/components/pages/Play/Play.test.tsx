import { Router } from 'react-router-dom';

import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { store } from 'store';

import { Play } from './Play';

const history = createMemoryHistory();
describe('Play', () => {
  beforeEach(() => {
    window.scrollTo = jest.fn();
  });

  it('Playのスナップショット', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <Play />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
