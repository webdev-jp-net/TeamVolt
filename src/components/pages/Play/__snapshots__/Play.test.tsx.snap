// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play Play„ÅÆ„Çπ„Éä„ÉÉ„Éó„Ç∑„Éß„ÉÉ„Éà 1`] = `
<DocumentFragment>
  <div
    class="l-page play"
  >
    <div
      class="visualizer"
    >
      <div
        class="stock"
      />
    </div>
    <div
      class="header"
    >
      <div
        class="timeLeftUi "
      >
        <svg
          class="gauge"
          version="1.1"
          viewBox="0 0 66 66"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M33,0c-0.6,0-1,0.4-1,1v4c0,0.6,0.4,1,1,1s1-0.4,1-1V1C34,0.4,33.6,0,33,0z M12.5,14.1c0.4,0.4,1,0.4,1.4,0 c0.4-0.4,0.4-1,0-1.4l-2.8-2.8c-0.4-0.4-1-0.4-1.4,0c-0.4,0.4-0.4,1,0,1.4L12.5,14.1z M54.5,9.9l-2.8,2.8c-0.4,0.4-0.4,1,0,1.4 c0.4,0.4,1,0.4,1.4,0l2.8-2.8c0.4-0.4,0.4-1,0-1.4C55.6,9.5,54.9,9.5,54.5,9.9z M12.5,51.9l-2.8,2.8c-0.4,0.4-0.4,1,0,1.4 c0.4,0.4,1,0.4,1.4,0l2.8-2.8c0.4-0.4,0.4-1,0-1.4C13.6,51.5,12.9,51.5,12.5,51.9z M53.1,51.9c-0.4-0.4-1-0.4-1.4,0 c-0.4,0.4-0.4,1,0,1.4l2.8,2.8c0.4,0.4,1,0.4,1.4,0c0.4-0.4,0.4-1,0-1.4L53.1,51.9z M33,60c-0.6,0-1,0.4-1,1v4c0,0.6,0.4,1,1,1 s1-0.4,1-1v-4C34,60.4,33.6,60,33,60z M65,32h-4c-0.6,0-1,0.4-1,1s0.4,1,1,1h4c0.6,0,1-0.4,1-1S65.6,32,65,32z M5,32H1 c-0.6,0-1,0.4-1,1s0.4,1,1,1h4c0.6,0,1-0.4,1-1S5.6,32,5,32z"
            fill="#ffffff"
          />
        </svg>
        <div
          class="progress"
        >
          <span
            class="count"
          >
            60
          </span>
        </div>
      </div>
      <p
        class="score"
      >
        üí• 0
      </p>
      <button
        class="tap"
        type="button"
      >
        Tap Screen!!
      </button>
    </div>
  </div>
</DocumentFragment>
`;
